prism-agent/service/wallet-api/src/main/scala/io/iohk/atala/agent/walletapi/vault/VaultClient.scala

```diff
@@ -41,6 +45,70 @@ class VaultKVClientImpl(vault: Vault) extends VaultKVClient {
     } yield ()
   }
 
+}
+
+object VaultKVClientImpl {
+
+  private final case class TokenRefreshState(leaseDuration: Duration)
+
+  def fromToken(address: String, token: String): Task[VaultKVClient] =
+    for {
+      vault <- createVaultInstance(address, Some(token))
+      vaultRef <- Ref.make(vault)
+    } yield VaultKVClientImpl(vaultRef)
+
+  def fromAppRole(address: String, roleId: String, secretId: String): Task[VaultKVClient] =
+    for {
+      vault <- createVaultInstance(address)
+      vaultRef <- vaultTokenRefreshLogic(vault, address, roleId, secretId)
+    } yield VaultKVClientImpl(vaultRef)
+
+  private def vaultTokenRefreshLogic(
```

Comment:
are we refreshing the token every 15 seconds?
would it be better to handle the invalid token error and update the token on the go?

Reply:
It refreshes the token 15 seconds before the current token expires. The token here is expected to be long-lived (default approle token setting is 1 month).